### PR TITLE
Sync v2 1 2 stable

### PR DESCRIPTION
**Pull request by jhtimmins**

### CODE REVIEW

**Review by ashb — APPROVED** [review]
Reminder to committers: DO NOT MERGE VIA GITHUB.

**Comment by ashb** [comment]
Oh, static checks failing cos 2.1 still has pylint.

Probably worth adding a few `# pylint: skip-file` to various places

**Comment by ashb** [comment]
I'll do that on v2-1-test directly.

**Review by kaxil — APPROVED** [review]

**Pull request merged**
